### PR TITLE
Topic 3

### PR DESCRIPTION
**Pull request by MaxiUlibarrie**
Hello. I made the api spec for the shopping cart. I have to do the rest of the topic 3 but I've wanted to know if I going on the right way. Thanks!

### CODE REVIEW

**Comment by MaxiUlibarrie** [comment]
I uploaded the rest of the topic 3.

File: topic-3/app-topic3/src/main/java/com/bootcamp/topic3/users/UserService.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package com.bootcamp.topic3.users;
+
+public interface UserService {
+
+  /**
+   * Creates a new user.
+   * @param user
+   */
+  public void createUser(User user);
```
[the remark this answers is not in the document]
Comment (MaxiUlibarrie) [reply topic-3/app-topic3/src/main/java/com/bootcamp/topic3/users/UserService.java]: Ok, I removed public modifiers in interface methods

File: topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/Item.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+   */
+  public Product getProduct() {
+    return product;
+  }
+
+  /**
+   * Replaces the product with a given one.
+   * @param product
+   */
+  public void setProduct(Product product) {
```
Comment (sebastianscatularo) [inline topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/Item.java]: usually, a good practice is trying to keep your classes immutables, that has some advantages 

https://en.proft.me/2013/11/24/advantages-and-disadvantages-immutable-objects-jav/
https://www.linkedin.com/pulse/20140528113353-16837833-6-benefits-of-programming-with-immutable-objects-in-java/

File: topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/ShopCart.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  /**
+   * If the given product exists in the cart
+   * then increases the product's amount, else
+   * creates a new item with amount 1.
+   * @param product given product
+   * @return true if the operation was successful
+   */
+  public boolean addToCart(Product product) {
+    for (Item item : cart) {
+      if (item.getProduct().equals(product)) {
```
Comment (sebastianscatularo) [inline topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/ShopCart.java]: seems this logic is similar to line 57, maybe you could refactor it

File: topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/ShopCart.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+   */
+  public void clearCart() {
+    cart.clear();
+  }
+
+  /**
+   * Gets checkout of all items from the cart.
+   * @return total price
+   */
+  public double checkOut() {
```
Comment (sebastianscatularo) [inline topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/ShopCart.java]: seems your checkout method doesn't do the right, you are returning the amount of all the cart, and you should mark the card as checked out

Comment (sebastianscatularo) [reply topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/ShopCart.java]: maybe mutating the state of the cart object

File: topic-3/shopping-cart-api-spec.yaml [hunk 5]
```diff
@@ -0,0 +1,217 @@
+swagger: "2.0"
```
Comment (sebastianscatularo) [inline topic-3/shopping-cart-api-spec.yaml]: nice!

File: topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/Item.java (hunk 2 above)
Comment (MaxiUlibarrie) [reply topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/Item.java]: Ok, I fixed that. Thanks!

File: topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/ShopCart.java (hunk 3 above)
Comment (MaxiUlibarrie) [reply topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/ShopCart.java]: Yes, you're right. I think I refactored that.

File: topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/ShopCart.java (hunk 4 above)
Comment (MaxiUlibarrie) [reply topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/ShopCart.java]: Ok, I added a state in ShopCart called checkedOut.

File: topic-3/shopping-cart-api-spec.yaml (hunk 5 above)
Comment (MaxiUlibarrie) [reply topic-3/shopping-cart-api-spec.yaml]: Thanks! 👍

File: topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/Item.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
+package com.bootcamp.topic3.shoppingcart;
+
+import java.util.Objects;
+
+/**
+ * Stores the product which the visitor
+ * want to buy and the solicited amount.
+ */
+public class Item {
```
[the remark this answers is not in the document]
Comment (MaxiUlibarrie) [reply topic-3/app-topic3/src/main/java/com/bootcamp/topic3/shoppingcart/Item.java]: Ok, I changed that in topic 6.

File: topic-3/app-topic3/src/test/java/com/bootcamp/topic3/shoppingcart/ShopCartTest.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    product2 = new Product(2,"EReader Y",120);
+    shopCart.addToCart(product1);
+    shopCart.addToCart(product2);
+  }
+
+  @Test
+  public void whenAddProductInCartThenItIsExistsInTheCartAndNotDuplicated() {
+    shopCart.addToCart(product1);
+
+    int duplicated = 0;
```
Comment (EduardoUTN) [inline topic-3/app-topic3/src/test/java/com/bootcamp/topic3/shoppingcart/ShopCartTest.java]: There's no need this for each block. You already know that List has product1 and product2 with 1 amount each. After the addToCart you only need to assert the amount equal 2 for that product. Maybe another assert checking the List size

**Pull request merged**